amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +153,224 @@ public void stop() {
 
     private void runShardSync() {
         if (leaderDecider.isLeader(workerId)) {
-            for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : currentStreamConfigMap.entrySet()) {
-                final ShardSyncTaskManager shardSyncTaskManager = shardSyncTaskManagerProvider.apply(streamConfigEntry.getValue());
-                if (!shardSyncTaskManager.syncShardAndLeaseInfo()) {
-                    log.warn("Failed to submit shard sync task for stream {}. This could be due to the previous shard sync task not finished.",
-                             shardSyncTaskManager.shardDetector().streamIdentifier().streamName());
+            log.info(String.format("WorkerId %s is leader, running the periodic shard sync task", workerId));
+            try {
+                // Construct the stream to leases map to be used in the lease sync
+                final Map<StreamIdentifier, List<Lease>> streamToLeasesMap = getStreamToLeasesMap(
+                        currentStreamConfigMap.keySet());
+
+                // For each of the stream, check if shard sync needs to be done based on the leases state.
+                for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : currentStreamConfigMap.entrySet()) {
+                    if (shouldDoShardSync(streamConfigEntry.getKey(),
+                            streamToLeasesMap.get(streamConfigEntry.getKey()))) {
+                        log.info("Periodic shard syncer initiating shard sync for {}", streamConfigEntry.getKey());
+                        final ShardSyncTaskManager shardSyncTaskManager = shardSyncTaskManagerProvider
+                                .apply(streamConfigEntry.getValue());
+                        if (!shardSyncTaskManager.castShardSyncTask()) {
+                            log.warn(
+                                    "Failed to submit shard sync task for stream {}. This could be due to the previous shard sync task not finished.",
+                                    shardSyncTaskManager.shardDetector().streamIdentifier().streamName());
+                        }
+                    } else {
+                        log.info("Skipping shard sync for {} as either hash ranges are complete in the lease table or leases hole confidence is not achieved.", streamConfigEntry.getKey());
```

Comment:
Can we make "lease hole confidence is not achieved" more explicit? E.g. "found {} consecutive holes, which does not meet confidence threshold of {} holes to trigger lease sync" or something similar?